src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DocumentPresentation/TextDocumentUriPresentationEndpoint.cs

```diff
@@ -0,0 +1,225 @@
+        private readonly DocumentResolver _documentResolver;
+        private readonly RazorDocumentMappingService _razorDocumentMappingService;
+        private readonly RazorComponentSearchEngine _razorComponentSearchEngine;
+        private readonly ILogger _logger;
+
+        public TextDocumentUriPresentationEndpoint(
+            ProjectSnapshotManagerDispatcher projectSnapshotManagerDispatcher,
+            DocumentResolver documentResolver,
+            RazorDocumentMappingService razorDocumentMappingService,
+            RazorComponentSearchEngine razorComponentSearchEngine,
+            ILoggerFactory loggerFactory)
+        {
+            if (projectSnapshotManagerDispatcher is null)
+            {
+                throw new ArgumentNullException(nameof(projectSnapshotManagerDispatcher));
+            }
+
+            if (documentResolver is null)
+            {
+                throw new ArgumentNullException(nameof(documentResolver));
+            }
+
+            if (razorDocumentMappingService is null)
+            {
+                throw new ArgumentNullException(nameof(razorDocumentMappingService));
+            }
+
+            if (razorComponentSearchEngine is null)
+            {
+                throw new ArgumentNullException(nameof(razorComponentSearchEngine));
+            }
+
+            if (loggerFactory is null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            _projectSnapshotManagerDispatcher = projectSnapshotManagerDispatcher;
+            _documentResolver = documentResolver;
+            _razorDocumentMappingService = razorDocumentMappingService;
+            _razorComponentSearchEngine = razorComponentSearchEngine;
+            _logger = loggerFactory.CreateLogger<TextDocumentUriPresentationEndpoint>();
+        }
+
+        public RegistrationExtensionResult GetRegistration()
+        {
+            const string AssociatedServerCapability = "_vs_uriPresentationProvider";
+
+            return new RegistrationExtensionResult(AssociatedServerCapability, true);
+        }
+
+        public async Task<WorkspaceEdit?> Handle(UriPresentationParams request, CancellationToken cancellationToken)
+        {
+            var documentSnapshot = await _projectSnapshotManagerDispatcher.RunOnDispatcherThreadAsync(() =>
+            {
+                _documentResolver.TryResolveDocument(request.TextDocument.Uri.GetAbsoluteOrUNCPath(), out var documentSnapshot);
+
+                return documentSnapshot;
+            }, cancellationToken).ConfigureAwait(false);
+
+            if (documentSnapshot is null)
+            {
+                _logger.LogWarning($"Failed to find document {request.TextDocument.Uri}.");
+                return null;
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var codeDocument = await documentSnapshot.GetGeneratedOutputAsync().ConfigureAwait(false);
```

Comment:
>GetGeneratedOutputAsync [](http://example.com/codeflow?start=54&length=23)

It seems a bit odd to call into a couple async methods without passing in the cancellation token

Reply:
Ya, we need to add a cancellation token to this, even if it doesn't consume it directly

Reply:
Yeah, I should have been better about this. In general Razor has lax usage of cancellation tokens, and obviously I copied this class from the wrong place :)

You know what they say though, the only thing for non-cancellable async code to prosper, is for good men to not add cancellation token parameters to thing.